src/libpriv/rpmostree-core.c

```diff
@@ -1715,6 +1739,88 @@ checkout_package_into_root (RpmOstreeContext *self,
   return TRUE;
 }
 
+static Header
+get_rpmdb_pkg_header (rpmts rpmdb_ts,
+                      DnfPackage *pkg,
+                      GCancellable *cancellable,
+                      GError **error)
+{
+  g_auto(rpmts) rpmdb_ts_owned = NULL;
+  if (!rpmdb_ts) /* allow callers to pass NULL */
+    rpmdb_ts = rpmdb_ts_owned = rpmtsCreate ();
+
+  unsigned int dbid = dnf_package_get_rpmdbid (pkg);
+  g_assert (dbid > 0);
+
+  g_auto(rpmdbMatchIterator) it =
+    rpmtsInitIterator (rpmdb_ts, RPMDBI_PACKAGES, &dbid, sizeof(dbid));
+
+  g_auto(Header) hdr = it ? rpmdbNextIterator (it) : NULL;
+  if (hdr == NULL)
+    return glnx_null_throw (error, "Failed to find package '%s' in rpmdb",
+                            dnf_package_get_nevra (pkg));
+
+  return headerLink (g_steal_pointer (&hdr));
+}
+
+static gboolean
+delete_package_from_root (RpmOstreeContext *self,
+                          rpmte         pkg,
+                          int           rootfs_dfd,
+                          GCancellable *cancellable,
+                          GError      **error)
+{
+  g_auto(rpmfiles) files = rpmteFiles (pkg);
+  g_auto(rpmfi) fi = rpmfilesIter (files, RPMFI_ITER_FWD);
+
+  g_autoptr(GPtrArray) deleted_dirs = g_ptr_array_new_with_free_func (g_free);
+
+  int i;
+  while ((i = rpmfiNext (fi)) >= 0)
+    {
+      /* see also apply_rpmfi_overrides() for a commented version of the loop */
+      const char *fn = rpmfiFN (fi);
+      rpm_mode_t mode = rpmfiFMode (fi);
+
+      if (!(S_ISREG (mode) ||
+            S_ISLNK (mode) ||
+            S_ISDIR (mode)))
+        continue;
+
+      g_assert (fn != NULL);
+      fn += strspn (fn, "/");
+      g_assert (fn[0]);
+
+      g_autofree char *fn_owned = NULL;
+      if (g_str_has_prefix (fn, "etc/"))
+        fn = fn_owned = g_strconcat ("usr/", fn, NULL);
+
+      /* for now, we only remove files from /usr */
+      if (!g_str_has_prefix (fn, "usr/"))
+        continue;
+
+      /* avoiding the stat syscall is worth a bit of userspace computation */
+      if (rpmostree_str_has_prefix_in_ptrarray (fn, deleted_dirs))
```

Comment:
Yeah...though at some point we should look for a radix tree implementation for this.